src/neuroglancer/ui/segment_select_tools.ts

```diff
@@ -0,0 +1,142 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {SegmentationUserLayer} from 'neuroglancer/segmentation_user_layer';
+import {removeChildren} from 'neuroglancer/util/dom';
+import {makeToolActivationStatusMessageWithHeader, registerLayerTool, Tool, ToolActivation} from 'neuroglancer/ui/tool';
+import { EventActionMap } from '../util/event_action_map';
+
+export const SELECT_SEGMENTS_TOOLS_ID = 'selectSegments';
+
+const selectEvent = 'mousedown0';
+const deselectEvent = 'mousedown2';
+const SELECT_SEGMENTS_INPUT_EVENT_MAP = EventActionMap.fromObject({
+  [`at:shift?+${selectEvent}`]: 'drag-select-segments',
+  [`at:shift?+${deselectEvent}`]: 'drag-deselect-segments',
+  'at:shift?+mouseup0': 'deactivate-drag-select-segments',
+  'at:shift?+mouseup2': 'deactivate-drag-deselect-segments',
+});
+
+enum ToolState {
+  IGNORE,
+  SELECT,
+  DESELECT,
+};
+
+export class SelectSegmentsTool extends Tool<SegmentationUserLayer> {
+  constructor(layer: SegmentationUserLayer, private state: ToolState) {
+    super(layer);
+  }
+
+  toJSON() {
+    return SELECT_SEGMENTS_TOOLS_ID;
+  }
+
+  activate(activation: ToolActivation<this>) {
+    const {layer} = this;
+    const {body, header} = makeToolActivationStatusMessageWithHeader(activation);
+    activation.bindInputEventMap(SELECT_SEGMENTS_INPUT_EVENT_MAP);
+    const updateStatus = () => {
+      removeChildren(body);
+      const msg = document.createElement('span');
+      switch (this.state) {
+        case ToolState.IGNORE:
+          header.textContent = 'Select/Deselect segments';
+          msg.textContent = `${selectEvent} to select segments; ${deselectEvent} to deselect segments.`;
+          break;
+        case ToolState.SELECT:
+        case ToolState.DESELECT:
+          header.textContent = `${this.state == ToolState.SELECT ? 'select' : 'deselect'} segments`;
+          msg.textContent = `Drag to ${this.state == ToolState.SELECT ? 'select' : 'deselect'} segments (${layer.displayState.segmentationGroupState.value.visibleSegments.size} selected).`;
+      }
+      body.appendChild(msg);
+    };
+    updateStatus();
+
+    const trySelectSegment = () => {
+      if (this.state == ToolState.IGNORE) {
+        return;
+      }
+      const {segmentSelectionState} = layer.displayState;
+      if (segmentSelectionState.hasSelectedSegment) {
+        const segment = segmentSelectionState.selectedSegment;
+        const {visibleSegments} = layer.displayState.segmentationGroupState.value;
+        switch (this.state) {
+          case ToolState.SELECT:
+            visibleSegments.add(segment);
+            break;
+          case ToolState.DESELECT:
+            if (visibleSegments.has(segment)) {
+              visibleSegments.delete(segment);
+            }
+        }
+      }
+    };
+
+    const startSelecting = () => {
+      updateStatus();
+      trySelectSegment();
+      activation.registerDisposer(
```

Comment:
Since this event handlers are never unregistered, `startSelecting` should not be called more than once per activation as it currently stands.